### PR TITLE
[mojo-stdlib] implement shifts in object

### PR DESCRIPTION
**Pull request by jiex-liu**
This is for implementing shift operation for object #2224

### CODE REVIEW

**Comment by LJ-9801** [comment]
@JoeLoser Hi, I have a questions in regards to all the bitwise operation for object. Python's behavior for shift and bitwise and/or/xor is different in that python will always cast the type of shift to an int where as the other bitwise operation will depend on the input types. 

In this pr I kind of just do a type check and then return a new object of type int for all the shift operator. Mean while I also found a lot of other bitwise operator such as and/or/xor that only support bool type
https://github.com/modularml/mojo/blob/abf2d97e88a2cc97e3029a578d2824895a88d0c1/stdlib/src/builtin/object.mojo#L1315
https://github.com/modularml/mojo/blob/abf2d97e88a2cc97e3029a578d2824895a88d0c1/stdlib/src/builtin/object.mojo#L1330

So I create another function name `_arithmetic_bitwise_op()` that handles these kind of operator for both int and bool. I did not use this function for the shift operation as it is not necessary, but it would be very helpful for completing the rest of the bitwise operator. 

Please let me know if it is OK to include that into this pr. The rest of the bitwise operator can be completed once this pr is merged. Thanks!

File: stdlib/src/builtin/object.mojo [hunk 1]
```diff
@@ -577,6 +577,18 @@ struct _ObjectImpl(CollectionElement, Stringable):
         else:
             lhs = lhs.convert_int_to_float()
 
+    @staticmethod
+    fn coerce_integral_type(inout lhs: _ObjectImpl, inout rhs: _ObjectImpl):
```
Comment (JoeLoser) [inline stdlib/src/builtin/object.mojo:581]: **Suggestion** Consider making this private (prefix with `_`).  Then, doc-gen won't generate public API docs for this as users shouldn't call this directly.

File: stdlib/src/builtin/object.mojo [hunk 2]
```diff
@@ -1198,6 +1216,30 @@ struct object(IntableRaising, Boolable, Stringable):
...
+            bool_func: Boolean operator.
+
+        Returns:
+            The bitwise operation result.
+        """
+        lhs._arithmetic_integral_type_check()
+        rhs._arithmetic_integral_type_check()
+        var lhsValue = lhs._value
+        var rhsValue = rhs._value
+        _ObjectImpl.coerce_integral_type(lhsValue, rhsValue)
```
Comment (JoeLoser) [inline stdlib/src/builtin/object.mojo:1238]: **Suggestion** Instead of `_ObjectImpl` you can use `Self` which is the same thing.

**Review by JoeLoser — APPROVED** [review]

**Comment by JoeLoser** [comment]
Looks good, thank you!  You'll need to rebase to resolve the test file conflict before we can merge.

**Comment by ConnorGray** [comment]
Hi @LJ-9801, thanks for this contribution! Joe and I talked about your question above re. bitwise operations. What you said about the other bitwise operators makes sense, but since it's separate from the problem of supporting shift operators, it would make review and discussion easier if we make those changes in a separate PR. This PR looks good as is though :)

Since `_arithmetic_bitwise_op` is not used in this PR, I suggest we leave it out of thise PR, and include it in a follow-up PR making the changes to the other bitwise operators you mentioned.

Thanks again for this contribution, very usefull! :fire:

**Comment by LJ-9801** [comment]
@ConnorGray Yeah sure! i will do that.

File: stdlib/src/builtin/object.mojo (hunk 1 above)
Comment (LJ-9801) [reply stdlib/src/builtin/object.mojo:586]: @JoeLoser  I implement this function inside the `_ObjectImpl` struct which I believe is private in nature right?

Comment (JoeLoser) [reply stdlib/src/builtin/object.mojo:586]: Oops, you're right here - sorry I missed that! 👍

**Comment by LJ-9801** [comment]
@JoeLoser @ConnorGray the function has been removed from this PR. I will open a issue for this as well as all the bitwise operation!

**Review by JoeLoser — APPROVED** [review]

**Comment by JoeLoser** [comment]
> @JoeLoser @ConnorGray the function has been removed from this PR. I will open a issue for this as well as all the bitwise operation!

Sounds good, feel free to do that.  I've just merged this PR.  Thanks again for the contribution! 🎉

**Pull request merged**